compiler/src/iree/compiler/Codegen/Common/ConvertToDestinationPassingStylePass.cpp

```diff
@@ -315,70 +315,153 @@ static bool canUseInOperandAsOutOperand(
   return true;
 }
 
-namespace {
-/// Adapts Linalg ops input operand to output operand. This is required for not
-/// creating extra alloca ops. For more details, see
-/// https://github.com/iree-org/iree/issues/8303
-struct AdaptLinalgInputOperandToOutputOperand
-    : public OpRewritePattern<linalg::GenericOp> {
-  AdaptLinalgInputOperandToOutputOperand(MLIRContext *context,
-                                         bool useWARForCooperativeMatrixCodegen,
-                                         PatternBenefit benefit = 1)
-      : OpRewritePattern(context, benefit),
-        useWARForCooperativeMatrixCodegen(useWARForCooperativeMatrixCodegen) {}
-
-  LogicalResult matchAndRewrite(linalg::GenericOp op,
-                                PatternRewriter &rewriter) const override {
-    // All the loops should be parallel loops.
-    if (op.getNumLoops() != op.getNumParallelLoops()) return failure();
-    // There is only one result tensor.
-    if (op->getNumResults() != 1) return failure();
-    // The output tensor is unused in the body computation.
-    auto outputOperand = op.getDpsInitOperand(0);
-    if (op.payloadUsesValueFromOperand(outputOperand)) return failure();
-
-    // Find an input operand which meets:
-    //   1. It has the same indexing map and type.
-    //   2. It is not from a readonly tensor.
-    OpOperand *operand = nullptr;
-    SmallVector<Value> newOperands;
-    SmallVector<AffineMap> maps;
-    for (auto in : op.getDpsInputOperands()) {
-      if (!operand &&
-          canUseInOperandAsOutOperand(in, outputOperand,
-                                      useWARForCooperativeMatrixCodegen)) {
-        operand = in;
-      } else {
-        newOperands.push_back(in->get());
-        maps.push_back(op.getMatchingIndexingMap(in));
-      }
-    }
-    if (!operand) return failure();
-    maps.push_back(op.getMatchingIndexingMap(operand));
+/// Checks if the use of a result of a compute op can be modified
+/// so that it can be moved into a store set.
+static std::optional<OpOperand *> canModifyUseToGetValueIntoStoreSet(
+    BufferizationPlan &plan, OpOperand *use,
+    bool useWARForCooperativeMatrixCodegen) {
+  assert(!plan.isInStoreSet(use->get()) &&
+         "attempting to move a value into a store set, when it is already part "
+         "of one");
+
+  // Currently only look at use in linalg.generic ops.
+  auto genericOpConsumer = dyn_cast<linalg::GenericOp>(use->getOwner());
+  if (!genericOpConsumer) return std::nullopt;
+
+  // All loops need to be parallel.
+  if (genericOpConsumer.getNumLoops() !=
+      genericOpConsumer.getNumParallelLoops())
+    return std::nullopt;
+
+  if (genericOpConsumer.isDpsInit(use)) return std::nullopt;
+
+  for (auto [index, outOperand] :
```

Comment:
nit: maybe we should rename `outOperand` to `initOperand`..

(It used to be out; it's changed to init after migrating to DPS interface.)